File: src/main/java/org/stt/text/JiraExpansionProvider.java
package org.stt.text;

import net.engio.mbassy.bus.MBassador;
import net.rcarz.jiraclient.Issue;
import org.stt.connector.jira.*;
import org.stt.event.NotifyUser;

import javax.inject.Inject;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.logging.Logger;

import static java.util.Objects.requireNonNull;

public class JiraExpansionProvider implements ExpansionProvider {

    private static final Logger LOG = Logger.getLogger(JiraExpansionProvider.class
            .getName());

    private final JiraConnector jiraConnector;
    private final Optional<MBassador<Object>> eventBus;

    @Inject
    public JiraExpansionProvider(JiraConnector connector,
                                 Optional<MBassador<Object>> eventBus) {
        this.jiraConnector = requireNonNull(connector);
        this.eventBus = eventBus;
    }

    @Override
    public List<String> getPossibleExpansions(String text) {
        String queryText = text.trim();
        int spaceIndex = queryText.lastIndexOf(' ');
        if (spaceIndex > 0) {
            queryText = text.substring(spaceIndex, text.length()).trim();
        }

        try {
            return jiraConnector.getIssue(queryText)
                    .map(Issue::getSummary)
                    .map(issue -> Collections.singletonList(": " + issue))
                    .orElse(Collections.emptyList());
        } catch (JiraConnectorException | AccessDeniedException | InvalidCredentialsException | IssueDoesNotExistException e) {
            eventBus.ifPresent(eb -> eb.publish(new NotifyUser(e.getMessage())));
            return Collections.emptyList();
        }
    }

}

File: src/main/kotlin/org/stt/text/JiraExpansionProvider.kt
package org.stt.text

import net.engio.mbassy.bus.MBassador
import org.stt.connector.jira.*
import org.stt.event.NotifyUser
import java.util.*
import javax.inject.Inject

class JiraExpansionProvider @Inject
constructor(private val jiraConnector: JiraConnector,
            private val eventBus: Optional<MBassador<Any>>) : ExpansionProvider {

    override fun getPossibleExpansions(text: String): List<String> {
        var queryText = text.trim { it <= ' ' }
        val spaceIndex = queryText.lastIndexOf(' ')
        if (spaceIndex > 0) {
            queryText = text.substring(spaceIndex, text.length).trim { it <= ' ' }
        }

        return try {
            jiraConnector.getIssue(queryText)?.summary?.let { listOf(": $it") } ?: emptyList()
        } catch (e: JiraConnectorException) {
            eventBus.ifPresent { eb -> eb.publish(NotifyUser(e.message ?: "")) }
            emptyList()
        } catch (e: Exceptions) {
            eventBus.ifPresent { eb -> eb.publish(NotifyUser(e.message ?: "")) }
            emptyList()
        } catch (e: InvalidCredentialsException) {
            eventBus.ifPresent { eb -> eb.publish(NotifyUser(e.message ?: "")) }
            emptyList()
        } catch (e: IssueDoesNotExistException) {
            eventBus.ifPresent { eb -> eb.publish(NotifyUser(e.message ?: "")) }
            emptyList()
        }

    }
}
